internals/daemon/daemon.go

```diff
@@ -203,18 +213,20 @@ func (c *Command) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	if rsp, ok := rsp.(*resp); ok {
 		st := c.d.state
-		st.Lock()
-		_, rst := restart.Pending(st)
-		st.Unlock()
-		switch rst {
-		case restart.RestartSystem:
-			rsp.transmitMaintenance(errorKindSystemRestart, "system is restarting")
-		case restart.RestartDaemon:
-			rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is restarting")
-		case restart.RestartSocket:
-			rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is stopping to wait for socket activation")
+		if includeMaintenance {
+			st.Lock()
+			_, rst := restart.Pending(st)
+			st.Unlock()
+			switch rst {
+			case restart.RestartSystem:
+				rsp.transmitMaintenance(errorKindSystemRestart, "system is restarting")
+			case restart.RestartDaemon:
+				rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is restarting")
+			case restart.RestartSocket:
+				rsp.transmitMaintenance(errorKindDaemonRestart, "daemon is stopping to wait for socket activation")
+			}
 		}
-		if rsp.Type != ResponseTypeError {
+		if includeWarningsSummary && rsp.Type != ResponseTypeError {
```

Comment:
This one is a little bit more tricky, because `WarningsSummary()` seems to depend on the current time:

https://github.com/canonical/pebble/blob/d822e9fd9c484a9c35e1b8f08b8d438ca2115225/internals/overlord/state/warning.go#L281-L297

Looking at the `addWarningsToMeta()` function that gets called with the result from `WarningsSummary()`:

https://github.com/canonical/pebble/blob/d822e9fd9c484a9c35e1b8f08b8d438ca2115225/internals/daemon/response.go#L68-L77

We can see that if `count == 0`, `addWarningsToMeta()` does nothing (also if `r.WarningCount != 0` already).

One thing that could avoid locks here is that if we have the `rsp` object (`Response` interface, but sometimes(?) implemented as type `resp` from `response.go`?), and its `WarningCount != 0`, we don't need to take out the lock and call `WarningsSummary()` (but that only applies to cases where `rsp` already contains warnings, not sure if that will be the case often - I expect not).

The other thing could be to - like above - track an atomic Uint64 (say) of how many warnings there are in state, not taking into account expiry. This would only work nicely if warnings in state are eventually purged if they are expired (otherwise as soon as one warning is in there, it won't ever go away).

One other way to track it would be the following:

- Keep track of a count of "unexpired warnings" and a "dirty flag" for that count
- Whenever a new warning is added, set the dirty flag
- Whenever the dirty flag is set here, call `WarningsSummary()` here, which will update the unexpired warnings count and reset the dirty flag
- Whenever the dirty flag is not set here, check "unexpired warnings" and if it's zero, skip the check

This can probably be simplified into a single atomic boolean flag "st.needToSummarizeWarningsWarnings":

- Adding a warning or changing any warnings in any way sets the flag
- `WarningsSummary()` returning `count == 0` resets the flag (otherwise the flag is kept set)
- The code here checks the flag, and only takes out the lock and calls `WarningsSummary()` when the flag is set

This assumes that warnings will never be visible "only in the future" (meaning at time `t` they are not visible, but at time `t + 1` they are visible), otherwise this logic would break.